lib/firmware/authenticator.c

```diff
@@ -0,0 +1,333 @@
+/*
+ * Copyright (C) 2022 markrypto
+ *
+ * TOPT generation as described in https://www.rfc-editor.org/rfc/rfc6238
+ * 
+ * This library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "keepkey/board/layout.h"
+#include "keepkey/board/confirm_sm.h"
+#include "keepkey/board/timer.h"
+#include "keepkey/firmware/authenticator.h"
+#include "keepkey/firmware/crypto.h"
+#include "keepkey/firmware/storage.h"
+#include "trezor/crypto/aes/aes.h"
+#include "trezor/crypto/base32.h"
+#include "trezor/crypto/hmac.h"
+#include "trezor/crypto/memzero.h"
+#include "trezor/crypto/sha2.h"
+
+#include "keepkey/firmware/app_confirm.h"
+#include "keepkey/firmware/app_layout.h"
+
+#include <string.h>
+#include <stdio.h>
+
+
+static CONFIDENTIAL authType authData[AUTHDATA_SIZE] = {0};
+
+
+// getAuthData() gets the storage version of authData and updates the local version
+// setAuthData() updates the storage version with the local version
+
+static void getAuthData(void) {
+  static bool localAuthdataUpdate = true; /* initialization trick */
+  if (localAuthdataUpdate) {
+    storage_getAuthData(authData);
+    localAuthdataUpdate = false;  
+  }
+  return;
+}
+
+static void setAuthData(void) {
+  storage_setAuthData(authData);
+}
+
+unsigned addAuthAccount(char *accountWithSeed) {
+  char *domain, *account, *seedStr;
+  unsigned slot;
+  char authSecret[20];          // 128-bit key len is the recommended minimum, this is room for 160-bit
+
+  // accountWithSeed should be of the form "domain:account:seedStr"
+  domain = strtok(accountWithSeed, ":");   // get the domain string token
+  if (NULL == domain) {
+    return 3;
```

Comment:
I think it will improve readability if we use #defines or enums for the error codes here.

Reply:
agreed